module/skill/Walk/src/Walk.cpp

```diff
@@ -0,0 +1,202 @@
+
+        // Main loop - Updates the walk engine at fixed frequency of UPDATE_FREQUENCY
+        on<Provide<WalkTask>,
+           Needs<LeftLegIK>,
+           Needs<RightLegIK>,
+           Every<UPDATE_FREQUENCY, Per<std::chrono::seconds>>,
+           Single>()
+            .then([this](const WalkTask& walk_task) {
+                switch (walk_engine.update(compute_time_delta(), walk_task.velocity_target.cast<double>()).value) {
+                    case WalkState::State::WALKING:
+                    case WalkState::State::STOPPING:
+                        walk();
+                        emit(std::make_unique<Stability>(Stability::DYNAMIC));
+                        break;
+                    case WalkState::State::STOPPED: emit(std::make_unique<Stability>(Stability::STANDING)); break;
+                    case WalkState::State::UNKNOWN:
+                    default: NUClear::log<NUClear::WARN>("Unknown state."); break;
+                }
+
+                // Emit the walking state
+                emit(std::make_unique<WalkState>(walk_engine.get_state(), walk_task.velocity_target));
+            });
+
+        // Stand Reaction - Sets walk_engine commands to zero, checks walk engine state and sets stability state
+        on<Provide<WalkTask>,
+           Needs<LeftLegIK>,
+           Needs<RightLegIK>,
+           Causing<Stability, Stability::STANDING>,
+           Every<UPDATE_FREQUENCY, Per<std::chrono::seconds>>,
+           Single>()
+            .then([this] {
+                // Stop the walk engine (request zero velocity)
+                switch (walk_engine.update(compute_time_delta(), Eigen::Vector3d::Zero()).value) {
+                    case WalkState::State::STOPPED: emit(std::make_unique<Stability>(Stability::STANDING)); break;
+                    case WalkState::State::STOPPING:
+                        walk();
+                        emit(std::make_unique<Stability>(Stability::DYNAMIC));
+                        break;
+                    case WalkState::State::WALKING: log<NUClear::WARN>("Walk engine state shouldn't be here."); break;
+                    case WalkState::State::UNKNOWN:
+                    default: NUClear::log<NUClear::WARN>("Unknown state"); break;
+                }
+
+                // Emit the walking state
+                emit(std::make_unique<WalkState>(walk_engine.get_state(), Eigen::Vector3f::Zero()));
+            });
+    }
+
+    double Walk::compute_time_delta() {
+        auto time_delta =
+            std::chrono::duration_cast<std::chrono::duration<double>>(NUClear::clock::now() - last_update_time).count();
+        last_update_time = NUClear::clock::now();
+        return time_delta;
+    }
+
+    void Walk::walk() {
+        // Compute the goal position time
+        const NUClear::clock::time_point time = NUClear::clock::now() + Per<std::chrono::seconds>(UPDATE_FREQUENCY);
```

Comment:
maybe rename this `goal_time`. I always try to avoid `time` cause it feels like it'll conflict way too easily, and with other libraries